src/main/java/org/springframework/social/partnercenter/security/AzureADMultiTenantOAuthOperations.java

```diff
@@ -0,0 +1,41 @@
+package org.springframework.social.partnercenter.security;
+
+import org.springframework.social.oauth2.AccessGrant;
+import org.springframework.social.oauth2.OAuth2Parameters;
+import org.springframework.social.partnercenter.http.logging.LogLevel;
+import org.springframework.util.MultiValueMap;
+
+public interface AzureADMultiTenantOAuthOperations {
+    String buildAuthorizeUrl(String clientId, String redirectUri, String state);
+
+    String buildAuthorizeUrl(String clientId, OAuth2Parameters additionalParams);
+
+    /**
+     * Exchange the authorization code for a refresh token
+     * @param authorizationCode the authorization code returned by the provider upon user authorization
+     * @param partnerTenantId the tenantId of the tenant that is being authorized
+     * @param redirectUri the authorization callback url; this value must match the redirectUri registered with the provider
+     * @param additionalParameters any additional parameters to be sent when exchanging the authorization code for an access grant. Should not be encoded.
+     * @return the access grant.
+     */
+    String exchangeForRefresh(String authorizationCode, String partnerTenantId, String redirectUri, MultiValueMap<String, String> additionalParameters);
+
+    MicrosoftRedirectAccessGrant extractRedirectAccessGrant(String body);
+
+    /**
+     * Exchange the refresh token for an access token
+     * @param refreshToken the refresh token returned by {@link #exchangeForRefresh}
+     * @param partnerTenantId the tenantId of the tenant that is being authorized
+     * @param resource the resource for the access grant
+     * @return the access grant.
+     */
+    AccessGrant exchangeForAccess(String refreshToken, String resource, String partnerTenantId);
```

Comment:
when we pass the refresh token it's usually refreshAccess. I understand that this is a bit different but if you just look at the method signature without looking at the param names you would assume the refreshToken param was the auth_code. Maybe we should state exchangeRefreshTokenForAccess. Then again maybe not. What do you think?

Reply:
I agree